### ISSUE TITLE
Possible issue(s) with processing RT-Stop Count type data

### ISSUE THREAD

**Issue by MeganSylvia** [dincarnato/RNAFramework#69]
Hi Danny,

I have been comparing the Ding 2014 and Siegfried 2014 scoring methods on processing RT Stop Count data of RNAs for my research and as I was digging through the results I noticed some inconsistencies. Basically, to verify some things on my end, I began calculating reactivity using the Ding and Siegfried equations in R and noticed some inconsistencies between my self-calculations and what was calculated via your program. I wanted to bring these two to your attention:

1) In calculating adjusted coverage for RT Stop Count type data: to fully replicate your methodology, I looked to your perl script for rf-count to see how you calculated adjusted coverage from the provided bam files. I noticed you take information from the left most position column and the CIGAR string to determine these coverage values, so I did the same thing (or at least a very similar thing) and noticed that I was getting large differences in coverage. Confused, I decided to use your coverage only rf-count option. Now I know this does not provide adjusted coverage, but, overall, the values should be in the same ballpark, however, there were also large discrepancies between the RT stop count adjusted coverage and the coverage-only coverage. Looking back to the code (I am not as familiar with perl) I think I found what might be the issue: 
Lines 867 to 891 describe the RT-stop count mode of calculating coverage in the rf-count script. I noticed that in line 869, you at least subtract 2 from the original left most position of the bam file, I am assuming to: 1) adjust the index to be from 0 to n to match up with perl indexing formats and 2) to perform the position adjustment to account for RT stops. However, I noticed that AFTER this indexing adjustment is applied (meaning any reads that read through the whole transcript will now have a left most position of -1), on line 871, a subsetting is applied to ignore all reads with a negative left most position. As a result of this, I think in only the coverage calculation (and not in the RT stop calling) the full length reads are accidentally ignored. I can provide my data files on this of course to help show you the discrepancy I noticed.

2) For this second inconsistency, I have not been able to pinpoint the exact section of code that may be the problem, but I thought I would mention it as well. As part of my analysis, I had used both the reactivity calculated from your program for the Ding 2014 scoring method with 2-8% normalization as well as reactivities that I had self-calculated in R AND reactivities that I had calculated using my lab's pipeline, StructureFold2. While I get identical RT stop counts called between the two programs and with my self-calculation in R, I am not getting identical values for either raw reactivity or 2-8% normalized reactivity. I am not sure what is causing the differences in values (again, I can provide the data I used to identify this). However, I have pinpointed one discrepancy that may be contributing to differences in values between the 2 programs, in regard to the 2-8% normalization: in the StructureFold2 pipeline, a 7 reactivity cap is employed, so no reactivity values can be > 7. In the SF2 pipeline, this capping is performed when the reactivity data is being exported (aka written to the final reactivity file). This means that if the user requests raw reactivity values (using the appropriate flag), the raw reactivity values are capped at 7 when being written to the final reactivity file, however, if the user specifies to have 2-8% normalized reactivity values, the raw reactivity values are not capped prior to the 2-8% normalization and instead the 2-8% normalized values are capped at 7 when written to the final reactivity file. I am not sure if the timing of capping truly matters and should not alter the values greatly, however, as mentioned above there is another discrepancy occurring between the two programs (using the same equation) that I have not been able to pinpoint that is altering the calculation of both raw and normalized reactivity. Note: when I self-calculated Ding 2014 reactivity in R, my values did agree with the SF2 calculated values and not those calculated with RNAFramework.

Please let me know if you have any questions about these descriptions, I tried to be as thorough as possible. Additionally, I did not attach the data files here yet, but I can easily provide the mapping/RT stop count/reactivity files that I used when I noticed these discrepancies! 

Thanks again,

Megan

**Comment by dincarnato** [comment]
Megan,

I seriously doubt that the subsetting of values with a negative index should cause such discrepancies. Actually, that subsetting is there just for safety and should almost never occur.

I suspect that the discrepancies are due to which reads are being included. For example, reads with 5' soft clippings, or reads with low mapping qualities are excluded.

To understand what is happening I will need the BAM file, reference, and counts you calculated for comparison.

Best,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny!

I really appreciate your help with this! I love the versatility of your pipeline! I am going to provide you with some of my data files, and please let me know if you need more (I just don't want to overwhelm you with files!)

So I had 3 biological replicates for each sample as well as 6 different sequencing runs performed on each of the biological replicates. I mapped my data to a tRNA rice reference. And I have plus DMS and minus DMS (control) data sets. Here I have attached:
1. The reference tRNA fasta file (Osativa_tRNA_CCA.fasta);
2. bam file for one sequencing run of one biological replicate of the Minus DMS (control) sample (CW2_1_1mm_filtered.bam);
3. bam file for one sequencing run of one biological replicate of the plus DMS sample (DW1_1_1mm_filtered.bam);
4. rc file generated using the RT Stop Count mode of rf-count for the minus DMS (control) bam file mentioned above (CW2_1_1mm_filtered.rc);
5. rc file generated using the RT Stop Count mode of rf-count for the plus DMS bam file mentioned above (DW1_1_1mm_filtered.rc);
6. rc file generated using the Coverage Only mode of rf-count for the minus DMS (control) bam file mentioned above (CW2_1_1mm_filtered_CovOnly.rc);
7. rc file generated using the Coverage Only mode of rf-count for the plus DMS bam file mentioned above (DW1_1_1mm_filtered_CovOnly.rc);
8. StructureFold2 RT stop count (rtsc) file in rtsc file format and a csv format for minus DMS (control) bam file mentioned above (CW2_1.rtsc; CW2_1_StructureFold2_RTStopCount.csv);
9. StructureFold2 RT stop count (rtsc) file in rtsc file format and a csv format for plus DMS bam file mentioned above (DW1_1.rtsc; DW1_1_StructureFold2_RTStopCount.csv);
10. Self calculated coverage file for the minus DMS (control) bam file mentioned above (CW2_1_SelfCalc_Coverage.csv);
11. Self calculated coverage file for the plus DMS bam file mentioned above (DW1_1_SelfCalc_Coverage.csv);
12. Self calculated rtsc file for the minus DMS (control) bam file mentioned above (CW2_1_SelfCalc_RTStopCount.csv);
13. Self calculated rtsc file for the plus DMS bam file mentioned above (DW1_1_SelfCalc_RTStopCount.csv);
14. RNAFramework reactivity file as csv: scoring method = Ding 2014; normalization method = none, raw reactivity; minus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; plus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; no coverage cutoff applied (DMS_RNAFramework_rawReact.csv)
15. RNAFramework reactivity file as csv: scoring method = Ding 2014; normalization method = 2-8% Normalization (reactive A and C nt normalized together); minus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; plus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; no coverage cutoff applied (DMS_RNAFramework_2-8NormReact.csv);
16. StructureFold2 reactivity file as csv: scoring method = Ding 2014; normalization method = none, raw reactivity; minus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; plus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; no coverage cutoff applied (DMS_StructureFold2_rawReact.csv);
17. StructureFold2 reactivity file as csv: scoring method = Ding 2014; normalization method = 2-8% Normalization (reactive A and C nt normalized together); minus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; plus sample = merged rc file of all three minus DMS (control) biological replicates and all six sequencing runs for each; no coverage cutoff applied (DMS_StructureFold2_2-8NormReact.csv).

Please let me know if there is anything else you need from me, but I think this should allow you to see the patterns I was mentioning above! 

I look forward to hearing back from you,

Thanks again!

Megan
[MS_Data.zip](https://github.com/user-attachments/files/16875397/MS_Data.zip)

**Comment by dincarnato** [comment]
Dear Megan,

Concerning issue (1), I have checked and I suspect you are using an outdated version of the RNA Framework.
When I try to calculate the coverage in RT-stop mode vs. coverage-only modes I get comparable coverage, while if I look in your files, the coverage in the RT-stop mode are significantly lower. This is possibly an issue that was already fixed quite some time ago... please try to `git pull` the latest version and try again.

__Note:__ I will be releasing the newest RNA Framework version on Monday

Concerning issue (2), I need a working test case. You refer to merged files in your description which, however, are not included in the zip, so I cannot replicate it and understand what is going on. Also, what you mentioned is not correct. The capping is not applied to the reactivities reported on file, but rather on the raw reactivities prior to normalization (see line 450 of `rf-norm`). 

So, this is what rf-norm does:
- Take the natural log of the raw RT count (in the ctrl or in the dms sample) + the pseudocount, at each position
- At each position, divide this number by the average of all the positions in that sample. According to the description provided in Ding *et al.*, 2014, the average is calculated over all positions of the RNA, not just A/C bases. *Maybe this is what you are doing differently?*
- At each position, the raw reactivity is calculated as *min(max(dms - ctrl, 0), reactivity_cap)*

Please let me know if you spot the difference.

Best,
Danny

**Comment by dincarnato** [comment]
Hi Megan,

the latest RNA Framework is now available.
Please issue a `git pull` and let me know if you get the right coverage now.

Cheers,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny,

Sorry for the late reply. To my knowledge I have the newest version of RNAFramework installed until your most recent update on Monday (rf-count v2.8.8), however I am running the update and will let you know if I see corrected values for the adjusted coverage! 

As for issue 2, I can provide additional data files. What exactly do you need from me to do this test case? The bam files? Or merged rc files? Just let me know and I can provide. 

Additionally, sorry for any confusion, but when I was referring to capping being applied to reactivity files when the reactivity data is exported, I was referring to when the capping is applied in the StructureFold2 pipeline. I agree this is different than when capping is applied in your pipeline (which is always after the calculation of raw reactivity). In the StructureFold2 pipeline, it is always applied to the reactivity when it is written to the final reactivity file, meaning if the user specifies they want raw reactivity, the raw reactivity is capped, but if the user specifies they want 2-8% normalized reactivity, then the 2-8% reactivity values are capped and not the raw reactivity values. I was simply stating this is one difference between the two pipelines, but I do not believe it is the only cause for the difference in values I observed between the StructureFold2 and RNAFramework calculations. 

In my investigations thus far, I have three different data sets: reactivity calculated by RNAFramework (using the Ding 2014 scoring method), reactivity calculated by StructureFold2 (which only used the Ding 2014 scoring method) and my self-calculated reactivity in R using the Ding 2014 equation. The RT stop count data between my R calculation, RNAFramework and StructureFold2 all agree perfectly with eachother. However, when raw reactivity is calculated, my R calculated values agree perfectly with those calculated via StructureFold2, but do not agree perfectly with the RNAFramework values (though the values are very close to each other). This difference in values increases if 2-8% normalization is performed. I have not been able to pinpoint the difference in the calculations, but I do know that I have included all nt, and not just A and C positions (because initially that is a mistake I made in my first attempt at reactivity calculations in R).

Thank you for all your help on this,

Megan

**Comment by dincarnato** [comment]
Dear Megan,

doesn't the reactivity capping alone explain the differences? What would happen if you were to cap your reactivities before normalization?
That said, I need an example to work on, so a specific transcript, and the calculated reactivities by StructureFold2, as well as an RC file with exactly the same count you used to derive your reactivities.

Best,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny, 
The capping alone does not explain the difference, because there is a difference between the output from the two programs when both raw reactivity is exported AND when 2-8% normalized reactivity is exported. I am going to attach RT stop count data, raw reactivity data, and 2-8% normalized reactivity data for a couple short transcripts (these files will have the data for the combined sequencing runs and biological replicates). I have attached these as a csv file that has all the necessary data in different columns and also as raw RNAFramework xml reactivity files and StructureFold2 .react files.

Please let me know if you have any questions.

And thanks again! 

Megan

P.S. the reference for this is the same as the reference provided previously.
[MS_Data2.zip](https://github.com/user-attachments/files/17018183/MS_Data2.zip)

**Comment by dincarnato** [comment]
Megan,

the information you sent me is still incomplete. In the data table there is no coverage column. Also, it takes me time to convert the data from one file to another. Since you have the RC file with those counts, can you please just share the RC file that will allow me replicate the issue?

Thanks.
Danny

**Comment by MeganSylvia** [comment]
Hi Danny!

Sorry about that! Here are the rc files for those data!
[MS_Data3.zip](https://github.com/user-attachments/files/17030405/MS_Data3.zip)


Megan

**Comment by dincarnato** [comment]
Hi Megan.

I am testing this on a single transcript, `EPlORYSAT000373814`.
Among other things, I tried to simply take the values you gave me from the `Data_for_troubleshooting.csv`, and input them into an Excel sheet. I then performed for each column one step of the raw reactivity calculation process:

1. take natural log of reactivities +1 in ctrl and dms
2. calculate average of log values in ctrl and dms
3. divide each log value by the average
4. calculate the difference

The values I obtain match the ones I get from RNA Framework. Can you please check the sheet and tell me what you or StructureFold2 are doing differently?

Cheers,
Danny

[raw_react.xlsx](https://github.com/user-attachments/files/17030520/raw_react.xlsx)

**Comment by MeganSylvia** [comment]
Hi Danny,

One piece that I do know is different from the calculations you performed here for raw reactivity from what is done in StructureFold2 (the updated version of computational pipeline discussed in the Ding 2014 paper) and what I did in R, is that you need to include the stops that ran off the 5' end of the transcript. 

Megan

**Comment by dincarnato** [comment]
I can try that, but those counts are not in the table you shared, are they? 
How do you count the RT stops past the 5'end if they would not count towards any base? Or do you assign them to the first base? In such a case, aren't they already in the Data table you shared? I need more context here...

**Comment by MeganSylvia** [comment]
Hi Danny,

Here are my self-calculated RT stop counts for the rc and rtsc files I provided previously. I obtained these by simply keeping the "0" position in the dataset after I performed the position adjustment (nt position - 1) to identify the modified nt position one position 5' of the RT stop. The 5' run off counts were not provided in the previous file, only because they are not included in the RT stop count output for either StructureFold2 or RNAFramework.  I am not the person who coded the StructureFold2 pipeline and python and/or perl coding is definitely not my expertise (I primarily use bash and R) so I am not sure how this information is carried through in the StructureFold2 scripts, but I do know that it is used, because my self-calculation in R validates it, as does the written text provided in Ding 2014 and the StructureFold2 manual.

I have also included two sam files here that contain the same data as the bam files provided above minus the header lines. I have provided these, because this is the input required by the StructureFold2 pipeline for the sam_to_rtsc.py script (in addition to the fasta file I provided previously). This would allow you to run the two pipelines on the same data set, just in case that helps you out! 

Let me know if you need anything else!

Thanks again!

Megan
[MS_Data4.zip](https://github.com/user-attachments/files/17033871/MS_Data4.zip)

**Comment by dincarnato** [comment]
Hi Megan,

so, did a test and it seems like this mostly explains the discrepancy, though the values are still slightly different (only by a few decimal points) from the one you calculated.

Do you see any other difference in the Excel sheet that could explain? Anyway, most of the difference seems to come from these very high counts deriving from full-length cDNAs. 

I am still not sure whether it is more correct to include them in the normalization or not. The reactivity profiles look basically identical between the two methods, the differences are marginal. The real question is, does the structure prediction differ between reactivities derived using SructureFold2 and those derived using RNA Framework?

I will see if I can add an option to consider these counts.

Best,
Danny

**Comment by dincarnato** [comment]
Megan,

I am trying to implement this, but the SAM files you gave me do not recapitulate the counts in your count tables for transcript `EPlORYSAT000373814` so I cannot test this.
Also, files are malformed (missing header).

I understand you do not want to share your data, but without an actual test case I am unable to test this. What I need is the actual SAM/BAM file, with just the `EPlORYSAT0003738141` transcript, with **ALL** the reads that led to the counts in the data table you originally shared.

Best,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny,

To make sure I answer all your points, I included your text in italics before my responses.

_so, did a test and it seems like this mostly explains the discrepancy, though the values are still slightly different (only by a few decimal points) from the one you calculated._

To my understanding, RNAFramework (with the Ding 2014 scoring method selected) and StructureFold2 should provide identical raw reactivity values if the same equation is being used to calculate both. Any small discrepancies in raw reactivity can be magnified when normalization is performed, increasing the discrepancy between the two methods.

_Do you see any other difference in the Excel sheet that could explain? Anyway, most of the difference seems to come from these very high counts deriving from full-length cDNAs._

I spent a long time trying to identify the exact cause of the discrepancies between raw reactivity calculated between RNAFramework and StructureFold2 prior to contacting you about this issue to make sure it was not something I was doing wrong on my end with either pipeline. I was unable to identify the mathematical cause for these discrepancies and therefore thought it was due to a piece of the code in either pipeline that I was misunderstanding. Though I was able to replicate identical StrucutureFold2 values in R, and unable to replicate the same exact RNAFramework values. Long story short, I am not sure what the difference is that explains this discrepancy in the raw reactivity values. 

_I am still not sure whether it is more correct to include them in the normalization or not. The reactivity profiles look basically identical between the two methods, the differences are marginal. The real question is, does the structure prediction differ between reactivities derived using SructureFold2 and those derived using RNA Framework?_

These values should be included in the initial scoring of reactivity (meaning the calculation of raw reactivity), but not in the chosen normalization calculation. In the normalization calculation, only the specified nucleotides (i.e. A and C for DMS) should be included. Since the 5' fall off reads have no nucleotide identity associated with them, they will not be included in the normalization. And yes they do cause differences in structure prediction. I get quite different prediction profiles for some of the included RNAs when I use RNAFramework Ding 2014 calculated reactivities and StructureFold2 calculated reactivities.

_I am trying to implement this, but the SAM files you gave me do not recapitulate the counts in your count tables for transcript [MS_Data5.zip](https://github.com/user-attachments/files/17049383/MS_Data5.zip) so I cannot test this.
Also, files are malformed (missing header)._
_I understand you do not want to share your data, but without an actual test case I am unable to test this. What I need is the actual SAM/BAM file, with just the EPlORYSAT0003738141 transcript, with ALL the reads that led to the counts in the data table you originally shared._

The sam files included in my last response were missing a header intentionally. They were identical to the bam files I originally provided in my first data share (that do have headers). I provided these sam files just in case you wanted to run the StructureFold2 pipeline directly, because the sam_to_rtsc.py script requires a sam file without header.

The sam/bam files I have were for each individual sequencing run/biological replicate, and the rc/rtsc/reactivity data was for the combination of all sequencing runs and bio reps, that is why the values do not agree with what was provided in the "troubleshooting" data set csv file. Sorry for any confusion. I usually collapse these after I generate the RT stop count type files so did not think about combining them before that step. I will include here a merged bam file for my minus and plus DMS samples for the transcript you are looking at! Hopefully that helps!

Again, thanks for looking into this,

Megan

**Comment by dincarnato** [comment]
Hi Megan,

thank you. I can almost completely recapitulate the raw reactivities now (only slight differences, but I think these have more to do with float handling and approximations). What I am really curious about is the 2-8% normalization. There seems to be some major difference there.

Can you please share your R code for that?
Here is what the RNA Framework does:

. Take the raw reactivities (only on A and C)
. Sort reactivities (higher to lower)
. Take the top 8% of values, after excluding the very top 2%
. Average those values and divide any other reactivity by that average

Cheers,
Danny

**Comment by dincarnato** [comment]
See another sheet to summarize the 2-8% normalization calculation. Please let me know what you do differently.

[norm.xlsx](https://github.com/user-attachments/files/17057791/norm.xlsx)

**Comment by dincarnato** [comment]
Hi Megan,

I figured out the difference.
Basically, both you and StructureFold, in case there are less than 50 values (with 50 values the top 2% would be the 1st value only), do not mark any value as outlier, and just use the top 8% to calculate the normalization factor.
However, this is incorrect. In such cases (so if 2% of the distribution is 0 values), the first value must be marked as outlier. This explains the final discrepancy.

I will be pushing the new rf-norm with the corrected score calculation, but I won't change the 2-8% normalization.

Best,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny!

Thank you for all your work on this and identifying that discrepancy in the 2-8% normalization! I really appreciate this! And I will also put in an issue note for StructureFold2 so this discrepancy with normalization of short reads can be fixed!

Megan

**Comment by dincarnato** [comment]
Hi Megan,

the new code has been pushed. Please issue a `git pull` and you should be good to go.

Cheers!
Danny

**Comment by MeganSylvia** [comment]
Yay! I'm excited!

**Comment by dincarnato** [comment]
Hi Megan, 

one last thing.
You might want to give a shot to `rf-normfactor`. The idea behind that module is to be able to calculate experiment-wide normalization factors, rather than normalizing each transcript separately. This is particularly effective when you have only short transcripts that are highly structured (so that will carry only a limited number of highly reactive bases). I am unsure whether all your sequences are short, but this might apply to your situation.

All you need to do is to run the module on your data, and then pass the resulting normalization factor to `rf-norm` through the `-nf` parameter.

I have just pushed the updated `rf-normfactor` module with the fixed Ding _et al._, 2014 scoring. Worth a shot :-)

Cheers,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny,

I appreciate that suggestion! Since my goal is to move away from short
transcripts for later analyses, I do not think I will be implementing it
now, but I did see that option in your pipeline. Have you ever thought of
expanding the rf-normfactor to allow for a list of normalization factors to
be inputted? That way if you have a lot of transcripts in your analysis,
you could apply the same normalization factor on the same transcript across
samples, but still allow for different normalization factors within sample.
Potentially useful for transcriptome wide analyses, especially where
expression varies greatly across transcripts!

Megan

**Comment by dincarnato** [comment]
Hi Megan,

not sure I follow. How can you use "a list of normalization factors"? If you have more than one normalization factor then it is not different than using rf-norm directly.
If by saying "you could apply the same normalization factor on the same transcript across samples" you mean using the same exact value across samples to divide by, that is not correct and not the way the "experiment-wise normalization" works.
When you do normalize within experiment, you do calculate all your raw reactivities as you normally would, across all the samples, then select only the bases that were covered in all experiments, and using those you identify your normalization factor per sample. 
Best,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny,

I described this idea poorly, I should have been more clear than simply
saying "samples." In my work, I look at how RNA structure, measured as
chemical probe reactivity, changes in response to an environmental stress,
such as heat or drought. So, when gathering this information, you would
have 4 "samples" total: a plus and minus chemical for the control growth
condition and a plus and minus chemical sample for the stressed growth
condition. When calculating reactivity using rf-norm, you would have to run
the script separately for the control growth condition and the experimental
growth condition. But what if you wanted to normalize the reactivities per
transcript by the same 2-8% normalization factor between the control and
stressed growth conditions? You could have normalization factors calculated
for the control growth condition normally (aka how rf-norm currently does
it), but instead of recalculating the normalization factor for the stressed
growth condition, you could simply divide by the normalization factors
previously calculated per transcript in the control condition.

Does that explain the idea more clearly?

Megan

**Comment by dincarnato** [comment]
Hi Megan,

I understand what you mean. You would like to have per-transcript normalization factors, calculated on the same set of bases for both experiments.
However. I believe that identifying a single normalization factor within each experiment (calculated on the same bases), and using it to normalize reactivities for all transcripts, is still more correct, because:

1. per-transcript normalization is very sensitive to the relative structuredness of each RNA (tightly folded RNAs might not have any, or very few, reactive bases... think for example of an extreme case in which you have a long stem of G-C pairs, with a UUUG loop on top, which you probe with DMS. All your Cs are actually base-paired, therefore lowly reactive, but the normalization does not take this into account, therefore the top 2-8% of reactivities (which are anyway low reactivities) will be "inflated" after the normalization and erroneously reported as highly reactive.
2. in the approach you proposed, let's hypothesize that the stress (e.g., cold) triggers a complete structural rearrangement, which results in the majority of the bases previously being single-stranded, now being double-stranded, and vice versa. You would not be able to find a normalization factor in this case as the bases in 2-8% of one condition will be different from those in the other condition.

If you instead calculate an experiment-wise distribution of reactivities, and only take the top 2-8% of those that are shared across the two experiments, now you have a normalization factor that is largely insensitive to the biases described above.

Best,
Danny

**Comment by dincarnato** [comment]
Megan,

I would like to flag that during the last update a bug was inadvertently introduced (2 lines commented out during tests and not uncommented before commit), which caused mis-calling of deletions in rf-count.
I apologize for this if it caused any inconvenience and advise you to pull the latest commit asap.

Moreover, we found out that on some systems the multithreading was having issues, causing, with files containing large numbers of transcripts (which I believe is your case), the number of running threads to decrease over time, down to just 1. This is now solved and, if you give it a shot, you should now observe a massive speed up in rf-count.

Cheers,
Danny

**Comment by MeganSylvia** [comment]
Hi Danny!

Thanks for this update! And I will definitely apply these updates ASAP!

I'm not sure if this was also noticed in your recent debugging over the
weekend, but I did notice that the "-" in some of my gene names (for
example: ENSRNA#######-T1) was getting replaced with an "_". I noticed this
in my rf-rctools view output, so I am not sure if it is just present in the
view output or happens at some other point in the pipeline.

Megan
